### PR TITLE
A new end point to get resource group stats

### PR DESCRIPTION
**Pull request by cemcayiroglu**
A load balancer can use resource group level stats to do better
load balancing. The new endpoint returns a ResourceGroupInfo.
ResourceGroupInfo contains a list of ResourceGroupInfo's as its subtrees.
ResourceGroupInfo consists of basic stats (running and queued query counts)
and actual queries as strings. This method leaves the queries empty
since it is not needed. The end method does not return the dynamically
generated the resource groups.

### CODE REVIEW

**Comment by mayankgarg1990** [comment]
A couple of high level questions:

1) Why is the name `fullStatsOnly` ?
2) `The end point returns ResourceGroupInfo that does not contain the running or queued queries` -> the resource group object does contain running and queued queries counts - not sure if I understood this line.
3) From an approach of returning all resource group info vs just one, have you thought about having a bulk status import? that way load balancer can cache this state as compared to making on demand calls per incoming query

**Comment by cemcayiroglu** [comment]
> A couple of high level questions:
> 
> 1. Why is the name `fullStatsOnly` ?

The end point returns all the stats for a resource group and all it's subtrees. I am open to name it differently. 

> 2. `The end point returns ResourceGroupInfo that does not contain the running or queued queries` -> the resource group object does contain running and queued queries counts - not sure if I understood this line.

The new endpoint returns a ResourceGroupInfo. ResourceGroupInfo contains a list of ResourceGroupInfo's as its subtrees. ResourceGroupInfo consists of basic stats (running and queued query counts) and actual queries as strings. This method leaves the queries empty since it is not needed.

> 3. From an approach of returning all resource group info vs just one, have you thought about having a bulk status import? that way load balancer can cache this state as compared to making on demand calls per incoming query

Good question.

The load balancer will call this method with a root group id. It will return everything under that resource group as a ResourceGroupInfo. I think this is the bulk status you are asking.   

CC: @mayankgarg1990

File: presto-main/src/main/java/com/facebook/presto/server/ResourceGroupStateInfoResource.java [hunk 1]
```diff
@@ -68,6 +68,27 @@ public ResourceGroupInfo getQueryStateInfos(@PathParam("resourceGroupId") String
         throw new WebApplicationException(NOT_FOUND);
     }
 
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    @Encoded
+    @Path("getStaticResourceGroupInfo/{resourceGroupId: .+}")
```
Comment (swapsmagic) [inline presto-main/src/main/java/com/facebook/presto/server/ResourceGroupStateInfoResource.java]: do we envision this endpoint to stay longer, if yes then i would say let's make it little little more generic, i.e. by taking a key=value as input where isStatic=true and use that instead of dedicating the entire endpoing to pull static resource group.

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java [hunk 2]
```diff
@@ -204,6 +213,30 @@ public ResourceGroupInfo getInfo()
         }
     }
 
+    public ResourceGroupInfo getStaticResourceGroupInfo()
+    {
+        synchronized (root) {
+            return new ResourceGroupInfo(
```
Comment (mayankgarg1990) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: Should we add a static resource group check invariant here. If somehow we make a call to this function and it is not static, then it should return a null (preferably an `Optional.empty()`) or raise an `UnsupportedOperationException` ?

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java [hunk 3]
```diff
@@ -87,6 +89,7 @@
     private final ResourceGroupId id;
     private final BiConsumer<InternalResourceGroup, Boolean> jmxExportListener;
     private final Executor executor;
+    private final boolean isStaticResourceGroup;
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: In my opinion the name is redundant, since the class is called `InternalResourceGroup`.  How about `isStatic`?

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java [hunk 4]
```diff
@@ -583,6 +621,11 @@ public InternalResourceGroup getOrCreateSubGroup(String name)
         }
     }
 
+    private boolean isSubResourceGroupStatic(Optional<ResourceGroupSelectionContext> resourceGroupSelectionContext)
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: Not sure I follow this.  Isn't this just checking if the present RG is static?  If so, why the name?

File: presto-spi/src/main/java/com/facebook/presto/spi/resourceGroups/ResourceGroupSelectionContext.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+    public static ResourceGroupSelectionContext createResourceGroupSelectionContext(int dynamicSegmentPosition)
+    {
+        return new ResourceGroupSelectionContext(OptionalInt.of(dynamicSegmentPosition));
+    }
+
+    public static ResourceGroupSelectionContext createResourceGroupSelectionContext(OptionalInt dynamicSegmentPosition)
+    {
+        return new ResourceGroupSelectionContext(dynamicSegmentPosition);
+    }
```
Comment (mayankgarg1990) [inline presto-spi/src/main/java/com/facebook/presto/spi/resourceGroups/ResourceGroupSelectionContext.java]: do we really need both of these? Should we just keep the `OptionalInt` one?

File: presto-main/src/main/java/com/facebook/presto/server/ResourceGroupStateInfoResource.java (hunk 1 above)
Comment (cemcayiroglu) [reply presto-main/src/main/java/com/facebook/presto/server/ResourceGroupStateInfoResource.java]: Makes sense I will change it. We should be able to receive the whole tree if needed.

File: presto-spi/src/main/java/com/facebook/presto/spi/resourceGroups/SelectionContext.java [hunk 6]
```diff
@@ -13,17 +13,28 @@
...
 {
     private final ResourceGroupId resourceGroupId;
     private final T context;
+    private final Optional<ResourceGroupSelectionContext> resourceGroupSelectionContext;
+
+    public SelectionContext(ResourceGroupId resourceGroupId, T context, ResourceGroupSelectionContext resourceGroupSelectionContext)
+    {
+        this.resourceGroupId = requireNonNull(resourceGroupId, "resourceGroupId is null");
+        this.context = requireNonNull(context, "context is null");
+        this.resourceGroupSelectionContext = requireNonNull(Optional.of(resourceGroupSelectionContext), "context is null");
```
Comment (mayankgarg1990) [inline presto-spi/src/main/java/com/facebook/presto/spi/resourceGroups/SelectionContext.java]: `Optional.of()` will never return a null. You probably want to change this to `Optional.of(requireNonNull())`

File: presto-spi/src/main/java/com/facebook/presto/spi/resourceGroups/SelectionContext.java [hunk 7]
```diff
@@ -13,17 +13,28 @@
...
 
+import java.util.Optional;
+
 import static java.util.Objects.requireNonNull;
 
 public final class SelectionContext<T>
 {
     private final ResourceGroupId resourceGroupId;
     private final T context;
+    private final Optional<ResourceGroupSelectionContext> resourceGroupSelectionContext;
```
Comment (mayankgarg1990) [inline presto-spi/src/main/java/com/facebook/presto/spi/resourceGroups/SelectionContext.java]: Can you better explain the decision to go for a separate class, rather than maybe just adding `dynamicFragmentId` here directly?

File: presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/ResourceGroupIdTemplate.java [hunk 8]
```diff
@@ -65,7 +69,11 @@ public ResourceGroupId expandTemplate(VariableMap context)
                 id = new ResourceGroupId(id, expanded);
             }
         }
-        return id;
+        OptionalInt first = IntStream.range(0, segments.size())
+                .filter(i -> segments.get(i).hasVariables())
+                .findFirst();
```
Comment (tdcmeehan) [inline presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/ResourceGroupIdTemplate.java]:
```suggestion
        OptionalInt first = segments.stream().filter(ResourceGroupNameTemplate::hasVariables).findFirst();
```

File: presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/ResourceGroupIdTemplate.java [hunk 9]
```diff
@@ -53,7 +57,7 @@ public static ResourceGroupIdTemplate fromSegments(List<ResourceGroupNameTemplat
         return new ResourceGroupIdTemplate(String.join(".", segments.stream().map(ResourceGroupNameTemplate::toString).collect(Collectors.toList())));
     }
 
-    public ResourceGroupId expandTemplate(VariableMap context)
+    public SelectionContext<VariableMap> expandTemplate(VariableMap context)
```
Comment (tdcmeehan) [inline presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/ResourceGroupIdTemplate.java]: It seems like we're changing the semantics of this method.  It's not really just expanding the template into a resource group id, it's giving you an entire selection context.  I would keep this method unchanged but make it private, and make a new method which calls this method to get the id, and also does the below (and it's public).

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java [hunk 10]
```diff
@@ -583,6 +621,11 @@ public InternalResourceGroup getOrCreateSubGroup(String name)
...
+    {
+        return !resourceGroupSelectionContext.equals(Optional.of(createResourceGroupSelectionContext(id.getSegments().size())));
```
Comment (swapsmagic) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: Coudn't understand the logic here, how are we deciding if a subgroup is static vs dynamic.  can you explain the logic here?

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java (hunk 3 above)
Comment (cemcayiroglu) [reply presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: Sure, I will change.

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java [hunk 11]
```diff
@@ -204,6 +213,30 @@ public ResourceGroupInfo getInfo()
         }
     }
 
+    public ResourceGroupInfo getStaticResourceGroupInfo()
```
[the remark this answers is not in the document]
Comment (cemcayiroglu) [reply presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: Sure, I am going to update.

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java (hunk 2 above)
Comment (cemcayiroglu) [reply presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: Makes sense. I am going to add it.

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java (hunk 4 above)
Comment (cemcayiroglu) [reply presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: This is actually checking if the sub resource groups is dynamic.

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java (hunk 10 above)
Comment (cemcayiroglu) [reply presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: resourceGroupSelectionContext is holding the dynamic segment index. 
For example:
abc_static.xyx_static.**qwe_dynamic**.sdq_static.
dynamic segment index is 2.
id.segment.size() is giving the size of the parent. This is also the index of the subgroup. so if parent.segment.size == dynamic segment index than the sub groups is dynamic.

File: presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/ResourceGroupIdTemplate.java (hunk 9 above)
Comment (cemcayiroglu) [reply presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/ResourceGroupIdTemplate.java]: I will change it.

File: presto-spi/src/main/java/com/facebook/presto/spi/resourceGroups/ResourceGroupSelectionContext.java (hunk 5 above)
Comment (cemcayiroglu) [reply presto-spi/src/main/java/com/facebook/presto/spi/resourceGroups/ResourceGroupSelectionContext.java]: I will keep the OptionalInt.

File: presto-spi/src/main/java/com/facebook/presto/spi/resourceGroups/SelectionContext.java (hunk 7 above)
Comment (cemcayiroglu) [reply presto-spi/src/main/java/com/facebook/presto/spi/resourceGroups/SelectionContext.java]: I was thinking that we could add more properties here. Swapnil had some plans. I can change it to dynamicFragmentId. Another option was adding this to the ResourceGroupId directly.

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java [hunk 12]
```diff
@@ -941,9 +987,12 @@ public int hashCode()
...
         private AtomicBoolean taskLimitExceeded = new AtomicBoolean();
 
-        public RootInternalResourceGroup(String name, BiConsumer<InternalResourceGroup, Boolean> jmxExportListener, Executor executor)
+        public RootInternalResourceGroup(
+                String name,
+                BiConsumer<InternalResourceGroup, Boolean> jmxExportListener,
+                Executor executor)
         {
-            super(Optional.empty(), name, jmxExportListener, executor);
+            super(Optional.empty(), name, jmxExportListener, executor, true);
```
Comment (shixuan-fan) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: Why is this `true`? Sorry I don't have much background about resource group.

File: presto-main/src/main/java/com/facebook/presto/server/ResourceGroupStateInfoResource.java [hunk 13]
```diff
@@ -68,6 +68,27 @@ public ResourceGroupInfo getQueryStateInfos(@PathParam("resourceGroupId") String
         throw new WebApplicationException(NOT_FOUND);
     }
 
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    @Encoded
+    @Path("getResourceGroupInfo/{resourceGroupId: .+}")
+    public ResourceGroupInfo getStaticResourceGroupInfo(@PathParam("resourceGroupId") String resourceGroupIdString)
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/server/ResourceGroupStateInfoResource.java]: question: Why do we need a separate method? It is technically very similar to the `getQueryStateInfos`. It feels like it is possible to simply add a parameter to the `getQueryStateInfos`. 

For example: 
  - `?includeQueryInfo=true|false`
  - `?summarizeSubgroups=true|false`
 - `?includeDynamicSubgroups=true|false`

Or maybe something like:

  - `?subgroupsInfo={SUMMARIZE|STATIC|ALL}`

File: presto-spi/src/main/java/com/facebook/presto/spi/resourceGroups/SelectionContext.java [hunk 14]
```diff
@@ -13,17 +13,28 @@
...
 
+import java.util.OptionalInt;
+
 import static java.util.Objects.requireNonNull;
 
 public final class SelectionContext<T>
 {
     private final ResourceGroupId resourceGroupId;
     private final T context;
+    private final OptionalInt dynamicSegmentPosition;
```
Comment (arhimondr) [inline presto-spi/src/main/java/com/facebook/presto/spi/resourceGroups/SelectionContext.java]: How about `firstDynamicSegmentPosition` (same for the getter)?

File: presto-main/src/main/java/com/facebook/presto/server/ResourceGroupInfo.java [hunk 15]
```diff
@@ -51,6 +52,7 @@
     private final List<ResourceGroupInfo> subGroups;
     private final List<QueryStateInfo> runningQueries;
 
+    @JsonCreator
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/server/ResourceGroupInfo.java]: Why is it needed?

If it is needed indeed - we should explicitly annotate all parameters of the constructor as we do for other Json serializable objects (e.g.: https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/execution/QueryInfo.java#L82)

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java [hunk 16]
```diff
@@ -87,6 +89,7 @@
     private final ResourceGroupId id;
     private final BiConsumer<InternalResourceGroup, Boolean> jmxExportListener;
     private final Executor executor;
+    private final boolean isStatic;
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: nit: `s/isStatic/static/`. By convention boolean variables in java are declared without `is`. Geter names for boolean variables are constructed as `is` + `VariableName`.

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java [hunk 17]
```diff
@@ -154,6 +162,7 @@ protected InternalResourceGroup(Optional<InternalResourceGroup> parent, String n
             id = new ResourceGroupId(name);
             root = this;
         }
+        this.isStatic = isStatic;
     }
 
     public ResourceGroupInfo getFullInfo()
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: How about changing this method signature to something like `public ResourceGroupInfo getFullInfo(boolean includeQueryInfo, boolean summarizeSubgroups, boolean includeDynamicSubgroups)`?

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java [hunk 18]
```diff
@@ -565,15 +604,22 @@ public void setJmxExport(boolean export)
         jmxExportListener.accept(this, export);
     }
 
-    public InternalResourceGroup getOrCreateSubGroup(String name)
+    public InternalResourceGroup getOrCreateSubGroup(String name, OptionalInt dynamicSegmentPosition)
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: This code is a little bit hard to follow. Could you please explain why is it needed to pass the `OptionalInt dynamicSegmentPosition` as a parameter? And why simple `boolean` is not enough? It feels like `createGroupIfNecessary` could be setting this boolean.

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java (hunk 12 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: I guess the root resource group is always static

Comment (cemcayiroglu) [reply presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: Yes!

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroupManager.java [hunk 19]
```diff
@@ -239,7 +240,9 @@ private synchronized void createGroupIfNecessary(SelectionContext<C> context, Ex
                 createGroupIfNecessary(new SelectionContext<>(id.getParent().get(), context.getContext()), executor);
                 InternalResourceGroup parent = groups.get(id.getParent().get());
                 requireNonNull(parent, "parent is null");
-                group = parent.getOrCreateSubGroup(id.getLastSegment());
+                // parent segments size equals to subgroup segment index
+                int subGroupSegmentIndex = parent.getId().getSegments().size();
+                group = parent.getOrCreateSubGroup(id.getLastSegment(), (context.getFirstDynamicSegmentPosition().equals(OptionalInt.of(subGroupSegmentIndex))));
```
Comment (swapsmagic) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroupManager.java]: i think this logic works to identify the given resource group is static or not. But a much cleaner way would be, if we mark each `segment` of `ResourceGroupId` being static or not when we create them via `group.expandTemplate` call, then we can easily put the flag here. And no need to rely on the context. 

i.e. this line will become something like this
group = parent.getOrCreateSubGroup(id.getLastSegment().getValue(), id.getLastSegment().isStatic());

And i think to make this happen, we may need to change the `segments` in `ResourceGroupId` class from string to a class Segment which looks like this:
class Segment {
    String value;
    String isStatic;
}

File: presto-main/src/main/java/com/facebook/presto/server/ResourceGroupStateInfoResource.java [hunk 20]
```diff
@@ -51,15 +53,22 @@ public ResourceGroupStateInfoResource(ResourceGroupManager<?> resourceGroupManag
     @Produces(MediaType.APPLICATION_JSON)
     @Encoded
     @Path("{resourceGroupId: .+}")
-    public ResourceGroupInfo getQueryStateInfos(@PathParam("resourceGroupId") String resourceGroupIdString)
+    public ResourceGroupInfo getQueryStateInfos(
+            @PathParam("resourceGroupId") String resourceGroupIdString,
+            @QueryParam("includeQueryInfo") @DefaultValue("true") boolean includeQueryInfo,
+            @QueryParam("summarizeSubgroups") @DefaultValue("true") boolean summarizeSubgroups,
+            @QueryParam("includeDynamicSubgroups") @DefaultValue("true") boolean includeDynamicSubgroups)
```
Comment (swapsmagic) [inline presto-main/src/main/java/com/facebook/presto/server/ResourceGroupStateInfoResource.java]: may be call this as `includeStaticSubgroupsOnly` to make it consistent with how we tag the resource group.

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroupManager.java (hunk 19 above)
Comment (cemcayiroglu) [reply presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroupManager.java]: We had a offline chat with @swapsmagic. We concluded that introducing a segment class would require a major code change. We are going to keep it as is.

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java [hunk 21]
```diff
@@ -87,6 +87,7 @@
     private final ResourceGroupId id;
     private final BiConsumer<InternalResourceGroup, Boolean> jmxExportListener;
     private final Executor executor;
+    private final boolean staticResourceGroup;
```
[the remark this answers is not in the document]
Comment (cemcayiroglu) [reply presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: static is a reserved key word.

[the remark this answers is not in the document]
Comment (swapsmagic) [reply presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: how about isStatic?

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java [hunk 22]
```diff
@@ -565,15 +578,22 @@ public void setJmxExport(boolean export)
         jmxExportListener.accept(this, export);
     }
 
-    public InternalResourceGroup getOrCreateSubGroup(String name)
+    public InternalResourceGroup getOrCreateSubGroup(String name, boolean staticSegment)
```
Comment (swapsmagic) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java]: i think this should have been named as `staticSubGroup`

File: presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroupManager.java [hunk 23]
```diff
@@ -103,10 +104,10 @@ public InternalResourceGroupManager(
     }
 
     @Override
-    public ResourceGroupInfo getResourceGroupInfo(ResourceGroupId id)
+    public ResourceGroupInfo getResourceGroupInfo(ResourceGroupId id, boolean includeQueryInfo, boolean summarizeSubgroups, boolean includeStaticSubgroupsOnly)
```
Comment (swapsmagic) [inline presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroupManager.java]: `summarizeSubgroups` should be `summarizeSubGroups` to make it true camel case.
